File: 0839_numSimilarGroups.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]

    def find(self, a):
        while a != self.parent[a]:
            a = self.parent[a]

        return a

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a == root_b:
            return

        self.parent[root_b] = root_a

class Solution:
    def numSimilarGroups(self, strs: List[str]) -> int:
        n = len(strs)
        uf = UnionFind(n)

        for i in range(n):
            for j in range(i + 1, n):
                if self.similar(strs[i], strs[j]):
                    uf.union(i, j)
                
        connected_components = 0

        for i in range(n):
            if i == uf.parent[i]:
                connected_components += 1
            
        return connected_components

    def similar(self, x, y):
        indices = []

        for i in range(len(x)):
            if x[i] != y[i]:
                indices.append(i)

            if len(indices) > 2:
                return False

        return True

File: 0501-1000/0839_numSimilarGroups.java
class UnionFind {
    int[] parent;
    int[] rank;

    public UnionFind(int n) {
        this.parent = new int[n];
        this.rank = new int[n];
        for (int i = 0; i < n; i++) this.parent[i] = i;
    }

    public int find(int a) {
        while (a != this.parent[a]) a = this.parent[a];
        return a;
    }

    public void union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return;
        if (this.rank[ra] > this.rank[rb]) this.parent[rb] = ra;
        else if (this.rank[ra] < this.rank[rb]) this.parent[ra] = rb;
        else {this.parent[rb] = ra; this.rank[ra]++;}
    }
}

class Solution {
    public int numSimilarGroups(String[] strs) {
        int n = strs.length;
        UnionFind uf = new UnionFind(n);
        HashSet<Integer> parents = new HashSet<>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isSimilar(strs[i], strs[j])) {
                    uf.union(i, j);
                }
            }
        }
        for (int i = 0; i < n; i++) parents.add(uf.find(i));
        return parents.size();
    }

    public boolean isSimilar(String a, String b) {
        int count = 0;
        for (int i = 0; i < a.length(); i++) if (a.charAt(i) != b.charAt(i)) count++;
        return count <= 2;
    }
}
